File: src/main/java/cn/lanink/autoresourcechest/chest/ChestManager.java
package cn.lanink.autoresourcechest.chest;

import cn.lanink.autoresourcechest.item.RandomItem;
import cn.nukkit.Server;
import cn.nukkit.item.Item;
import cn.nukkit.level.Position;
import cn.nukkit.utils.Config;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * @author lt_name
 */
@EqualsAndHashCode
@Getter
public class ChestManager {

    private final String name;

    private final String showName;
    private final int refreshInterval;
    private final int restrictOpenCount;
    private final int maxRandomItemCount;

    public Item[] fixedItems;
    public ArrayList<RandomItem> randomItems = new ArrayList<>();

    private final Map<Position, Chest> chests = new HashMap<>();

    public ChestManager(@NotNull String name, @NotNull Config config) {
        this.name = name;

        this.showName = config.getString("showName");
        this.refreshInterval = config.getInt("刷新间隔(s)");
        this.restrictOpenCount = config.getInt("限制打开次数", -1);
        this.maxRandomItemCount = config.getInt("生成随机物品数量限制");

        ArrayList<Item> list = new ArrayList<>();
        for (String stringItem : config.getStringList("fixedItem")) {
            String[] split = stringItem.split("&");
            Item item = Item.fromString(split[0]);
            item.setCount(Integer.parseInt(split[1]));
            list.add(item);
        }
        this.fixedItems = list.toArray(new Item[0]);

        for (String stringItem : config.getStringList("randomItem")) {
            this.randomItems.add(new RandomItem(stringItem));
        }

        for (String pos : config.getStringList("pos")) {
            String[] split = pos.split(":");
            Position position = new Position(Integer.parseInt(split[0]),
                    Integer.parseInt(split[1]),
                    Integer.parseInt(split[2]),
                    Server.getInstance().getLevelByName(split[3]));
            this.chests.put(position, new Chest(this, position));
        }

    }

    public Item[] getFixedItems() {
        return this.fixedItems.clone();
    }

    public Item[] getRandomItems() {
        ArrayList<Item> list = new ArrayList<>();
        for (RandomItem randomItem : this.randomItems) {
            Item item = randomItem.getItem();
            if (item != null) {
                list.add(item);
                if (list.size() >= this.getMaxRandomItemCount()) {
                    break;
                }
            }
        }
        return list.toArray(new Item[0]);
    }

}

File: src/main/java/cn/lanink/autoresourcechest/chest/ChestManager.kt
package cn.lanink.autoresourcechest.chest

import cn.lanink.autoresourcechest.item.RandomItem
import cn.nukkit.Server
import cn.nukkit.item.Item
import cn.nukkit.level.Position
import cn.nukkit.utils.Config
import lombok.EqualsAndHashCode

/**
 * @author lt_name
 */
@EqualsAndHashCode
class ChestManager(val name: String, private val config: Config) {

    val showName: String = config.getString("showName")
    val refreshInterval: Int = config.getInt("刷新间隔(s)")
    private val restrictOpenCount: Int = config.getInt("限制打开次数", -1)
    private val maxRandomItemCount: Int = config.getInt("生成随机物品数量限制")
    private var fixedItems: ArrayList<Item> = ArrayList()
    private var randomItems = ArrayList<RandomItem>()
    val chests: MutableMap<Position, Chest> = HashMap()

    init {
        for (stringItem in config.getStringList("fixedItem")) {
            val split = stringItem.split("&")
            val item = Item.fromString(split[0])
            item.setCount(split[1].toInt())
            this.fixedItems.add(item)
        }
        for (stringItem in config.getStringList("randomItem")) {
            this.randomItems.add(RandomItem(stringItem!!))
        }
        for (pos in config.getStringList("pos")) {
            val split = pos.split(":")
            val position = Position(
                split[0].toDouble(), split[1].toDouble(), split[2].toDouble(),
                Server.getInstance().getLevelByName(split[3])
            )
            this.chests[position] = Chest(this, position)
        }
    }

    fun saveConfig() {
        val list = mutableListOf<String>()
        for (pos in this.chests.keys) {
            list.add("${pos.x}:${pos.y}:${pos.z}:${pos.level.name}")
        }
        this.config.set("pos", list)
        this.config.save()
    }

    fun addNewChest(position: Position): Boolean {
        for (pos in this.chests.keys) {
            if (pos.getLevel() === position.getLevel() && pos == position) {
                return false
            }
        }
        val newPos = position.clone()
        this.chests[newPos] = Chest(this, newPos)
        return true
    }

    fun removeChest(chest: Chest): Boolean {
        for ((key, value) in HashMap(this.chests).entries) {
            if (value == chest) {
                this.chests.remove(key)
                value.close()
                return true
            }
        }
        return false
    }

    fun closeAllChest() {
        this.chests.values.forEach {
            chest -> chest.close()
        }
        this.chests.clear()
    }

    fun getFixedItems(): List<Item> {
        return ArrayList<Item>(this.fixedItems)
    }

    fun getRandomItems(): List<Item> {
        val list = ArrayList<Item>()
        for (randomItem in this.randomItems) {
            val item = randomItem.getItem()
            if (item != null) {
                list.add(item.clone())
                if (list.size >= this.maxRandomItemCount) {
                    break
                }
            }
        }
        return list
    }

    fun getChestByPos(position: Position): Chest? {
        //需要判断level
        for ((key, value) in this.chests.entries) {
            if (key.getLevel() === position.getLevel() && key == position) {
                return value
            }
        }
        return null
    }

}